spec/std/uri/params/from_www_form_spec.cr

```diff
@@ -0,0 +1,151 @@
+    end
+
+    it "a falsey value" do
+      Bool.from_www_form("false").should be_false
+      Bool.from_www_form("off").should be_false
+      Bool.from_www_form("no").should be_false
+      Bool.from_www_form("0").should be_false
+    end
+
+    it "any other value" do
+      Bool.from_www_form("foo").should be_nil
+      Bool.from_www_form("").should be_nil
+    end
+  end
+
+  describe String do
+    it "scalar string" do
+      String.from_www_form("John Doe").should eq "John Doe"
+    end
+
+    it "with key" do
+      String.from_www_form(URI::Params.new({"name" => ["John Doe"]}), "name").should eq "John Doe"
+    end
+
+    it "with missing key" do
+      String.from_www_form(URI::Params.new({"" => ["John Doe"]}), "name").should be_nil
+    end
+
+    it "with alternate casing" do
+      String.from_www_form(URI::Params.new({"Name" => ["John Doe"]}), "name").should be_nil
+    end
+
+    it "empty value" do
+      String.from_www_form(URI::Params.new({"name" => [""]}), "name").should eq ""
+    end
+  end
+
+  describe Enum do
+    it "valid value" do
+      Color.from_www_form("green").should eq Color::Green
+    end
+
+    it "invalid value" do
+      expect_raises ArgumentError do
+        Color.from_www_form ""
+      end
+    end
+  end
+
+  describe Time do
+    it "valid value" do
+      Time.from_www_form("2016-11-16T09:55:48-03:00").to_utc.should eq(Time.utc(2016, 11, 16, 12, 55, 48))
+      Time.from_www_form("2016-11-16T09:55:48-0300").to_utc.should eq(Time.utc(2016, 11, 16, 12, 55, 48))
+      Time.from_www_form("20161116T095548-03:00").to_utc.should eq(Time.utc(2016, 11, 16, 12, 55, 48))
+    end
+
+    it "invalid value" do
+      expect_raises Time::Format::Error do
+        Time.from_www_form("").should be_nil
+      end
+    end
+  end
+
+  describe Nil do
+    it "valid values" do
+      Nil.from_www_form("").should be_nil
+    end
+
+    it "invalid value" do
+      expect_raises ArgumentError do
+        Nil.from_www_form("null").should be_nil
+      end
+    end
+  end
+
+  describe Number do
+    describe Int do
+      it "valid numbers" do
+        Int64.from_www_form("123").should eq 123_i64
+        UInt8.from_www_form("7").should eq 7_u8
+        Int64.from_www_form("-12").should eq -12_i64
+      end
+
+      it "with whitespace" do
+        expect_raises ArgumentError do
+          Int32.from_www_form(" 123")
+        end
+      end
+
+      it "empty value" do
+        expect_raises ArgumentError do
+          Int16.from_www_form ""
+        end
+      end
+    end
+
+    describe Float do
+      it "valid numbers" do
+        Float32.from_www_form("123.0").should eq 123_f32
+        Float64.from_www_form("123.0").should eq 123_f64
+      end
+
+      it "with whitespace" do
+        expect_raises ArgumentError do
+          Float64.from_www_form(" 123.0")
+        end
+      end
+
+      it "empty value" do
+        expect_raises Exception do
+          Float64.from_www_form ""
+        end
```

Comment:
This might be worth it's own issue:

```cr
$ crystal eval 'pp Float64.new "", whitespace: false'
Unhandled exception: Index out of bounds (IndexError)
  from /usr/lib/crystal/string.cr:996:16 in '[]'
  from /usr/lib/crystal/string.cr:755:40 in 'to_f64?'
  from /usr/lib/crystal/string.cr:710:5 in 'to_f64'
  from /usr/lib/crystal/float.cr:327:5 in 'new:whitespace'
  from eval:1:1 in '__crystal_main'
  from /usr/lib/crystal/crystal/main.cr:118:5 in 'main_user_code'
  from /usr/lib/crystal/crystal/main.cr:104:7 in 'main'
  from /usr/lib/crystal/crystal/main.cr:130:3 in 'main'
  from /usr/lib/libc.so.6 in '??'
  from /usr/lib/libc.so.6 in '__libc_start_main'
  from /home/george/.cache/crystal/crystal-run-eval.tmp in '_start'
  from ???
```
I'd expect this to bubble up an `ArgumentError`. At the moment it seems to bubble up an exception from the implementation when using `whitespace: false`.